### ISSUE TITLE
Dark mode custom color hexcodes don't display accurately

### ISSUE THREAD

**Issue by pmnord** [excalidraw/excalidraw#3531]
I'm inputting custom color hex codes, but they only display correctly in light mode.

![image](https://user-images.githubusercontent.com/14935197/116929683-c6df6600-ac2c-11eb-8cf9-667f83aab696.png)

**Comment by dwelle** [comment]
Yeah, we just `invert` the colors and then `hue-rotate` them and go with whatever it gives us. Works pretty solid for the most part.

**Comment by a-silent-message** [comment]
Hi, how would i go by using a custom color as my background in dark mode then? since i dont know how much hue-rotation is done?

**Comment by Mobuos** [comment]
Any developments on this issue? I made a suggestion for keeping dark mode as is and unaffecting on-going projects while allowing hexcode-accurate colors in dark mode in https://github.com/excalidraw/excalidraw/issues/4616#issuecomment-1017028145, is that feasible?

**Comment by niv-dev** [comment]
This is the current solution that I have in place:

Use the [Stylus](https://chrome.google.com/webstore/detail/stylus/clngdbkpkpeebahjckkjfobafhncgmne) extension, and import / write the following for the domain excalidraw.com: 

```
.excalidraw.theme--dark {
    --theme-filter: invert(1)
}
```

This overrides the hue changes and thus simply inverts the color, the inverted color of any hex code can simply be found using online tools. So all in all, for my configuration which ends up looking like this (notice how the background is darker then what is allowed with the normal limitations): 

![image](https://user-images.githubusercontent.com/79858886/155151872-d0b3d4fc-ce15-4fff-a058-eb79c85fb77b.png). 

_tldr: use stylus, import the css mentioned, and to put in a custom hex code, simply find the chromatic inverse of that color (`000000 -> ffffff`)_.

**Comment by codercatdev** [comment]
Also if you don't want to install anything.

```
const theme = Array.from(document.getElementsByClassName('theme--dark')).pop();
theme.style.setProperty('--theme-filter', 'invert(0)' );
```

**Comment by Polina-Hapi** [comment]
> Also if you don't want to install anything.
> 
> ```
> const theme = Array.from(document.getElementsByClassName('theme--dark')).pop();
> theme.style.setProperty('--theme-filter', 'invert(0)' );
> ```

@codercatdev  Would you by any chance know how to do this modification in Obsidian? I added it as a new CSS file but it doesn't seem to make a difference. Only later I realised this thread is discussing the web version.

**Comment by u0nel** [comment]
> Yeah, we just invert the colors and then hue-rotate them and go with whatever it gives us. Works pretty solid for the most part.

Wouldn't it then make sense to also display the resulting color code in the color chooser and use the inverted color instead if custom colors are inserted?

**Comment by dwelle** [comment]
So the options are:

1. do not adapt (rotate/change) colors to dark mode at all

      **pros:**
     
      - color accuracy
      - control
    
      **cons:** 
    
      - requires user intervention. No more "it just works" (when it works). Many colors won't be anywhere close to usable, or be visible at all (e.g. black on black/dark bg).

2. mix of handpicking dark mode colors for the official color palette, and auto-adapting the rest.

      **pros:**
     
      - official color palette's color will be "guaranteed" to look good in dark mode
    
      **cons:** 

      - downsides of current behavior for custom colors
      - people who want color accuracy won't be satisfied

4. mix of handpicking dark mode colors for the official color palette, while leaving the rest unchanged.

      **pros:**
     
      - will work better™
    
      **cons:** 

      - unintuitive behavior (mix or auto-adapting color and keeping as is)
      - downsides of (1) for custom colors
      - people who want color accuracy won't be satisfied with the official color palette's dark mode colors

5. change the color-rotation algorithm so it's better™

      **pros:**
     
      - will work better™
    
      **cons:** 

      - people who want color accuracy won't be satisfied

6. add a toggle to switch the drawing (or specific elements) to either auto-adapting colors to dark mode, or leaving them unchanged (1)

      **pros:**
     
      - control
      - the default could stay to what we do now so that people who do not care about color accuracy won't need to do anything
    
      **cons:** 

      - (UI) complexity

7. when we switch to dark mode, auto-adapt the color, but also explicitly changing the `HEX` color (and the color in the UI), allowing to manually change to something else (via https://github.com/excalidraw/excalidraw/issues/4616#issuecomment-1017028145)

      **pros:**
     
      - control
      - works as is by default, but allows control
    
      **cons:** 

      - I'd slightly unintuitive behavior?

8. ???

**Comment by mart3323** [comment]
Would it be an option to have the conversion done in the color picker itself? 
So the color picker would take the light theme value from the diagram but show the dark theme hex instead - and when the user edits it, it would run the filters backwards to modify the underlying light theme value.

react pseudocode:
```tsx
const toDark = color => hueRotate(180, invert(0.95, color));
const fromDark = color => invert(1/0.95, hueRotate(180, color))

const ColorPicker = ({color, setColor}) => {
  const theme = useTheme()
  const [inputValue, setInputValue] = useState(null); // to delay rounding errors until the user is finished typing
  
  const percievedValue = theme === 'dark' ? toDark(color) : color
  return <Input 
    value={inputValue ?? percievedValue}
    onChange={v => {
      setInputValue(v);
      setColor(theme === 'dark' ? setColor(fromDark(v)) : v)
    })
    onBlur={() => setInputValue(null)}
  />
}
```
pros:
* No new UI required
* Existing diagrams are unaffected, switching themes is unaffected, dark theme rendering is unaffected

cons:
* not every color will be usable, due to rounding errors (`#c0ffee` -> `#??????` -> `#c1ffed`) or because the invert is 95% which clips out extreme lights and darks (`#ffffff` -> `#000000` -> `#ededed`)

**Comment by Zirafnik** [comment]
Honestly, as a new user, this was a very bad user experience today. My default theme for everything is dark. 

The hex colors I wanted to use just didn't display properly, until I realized, that the brighter the original hex, the darker the actually displayed color. It's inverse!

Now I understand that you need to adjust the color scheme if you are switching from light to dark, in order to keep the contrast, but please do not apply this feature to new colors you pick.

This definitely isn't the expected behavior. I propose that colors already on screen should be inversed, while new colors from the hex color picker should stay true to their value.

**Comment by mart3323** [comment]
> Would it be an option to have the conversion done in the color picker itself?

Had the chance to test this idea, and i didn't realise hue-rotate is also lossy (and much worse than the invert(92%)).
Basically all the vibrant colors are gone, so i would only take this as a last resort if none of the other options are workable.

<details><summary>Screenshots</summary>

Test diagram with no filters in light theme:
<img src="https://user-images.githubusercontent.com/8430090/229289181-bd0cc3bf-8696-4a7a-ac97-f42f7b2471ab.png" width="50%"/>
Test diagram in current dark theme
<img src="https://user-images.githubusercontent.com/8430090/229289928-0512a103-3647-4307-b6fd-175ab3fdb8b9.png" width="50%"/>
Custom implementation of ColorPicker that reverses the filters in js
<img src="https://user-images.githubusercontent.com/8430090/229289344-10a3558e-d679-4b93-9103-3d897879aed1.png" width="50%"/>
css-only test with both filters
<img src="https://user-images.githubusercontent.com/8430090/229289264-c17ac85e-b7fa-4582-9190-c123a5ef94e7.png" width="50%"/>
css-only test with only hue-rotate
<img src="https://user-images.githubusercontent.com/8430090/229289232-2d56fc85-f6a0-4231-a241-84c5e2efa8ec.png" width="50%"/>

</details>

**Comment by TebyTeb** [comment]
Can't we just have a color picker in-app for custom colors? So you can visually see the changes to the color we want to presonalize. No need of touching original palette, you can use it as groundbase to find your exact tone.

**Comment by inSilicum** [comment]
I reversed the process of inversion and hue-rotation and received #7f7f00 as a code that does paint a decent yellow in the dark mode in Excalidraw. One could fine-tune this hex code to compute the exact varietals of yellow one may be looking for.

**Comment by dwelle** [comment]
status: we're looking into solution (2) from https://github.com/excalidraw/excalidraw/issues/3531#issuecomment-1406777374, potentially mixed with some form of (6) down the line.

The problem with (2) though is that the dark theme palette will still not look anywhere near close to light theme colors for light colors, because we still need to ensure that the text, which gets inverted to white if black in light mode, is legible on the background if it was legible in light mode.

Example: (the colors are not final; 1-5 indicate color shades; shade 4 currently no being changed in dark mode)

![image](https://user-images.githubusercontent.com/5153846/231587153-4e870328-aca6-4970-8135-ac20bcac8599.png)

**Comment by eddie-knight** [comment]
I ran into this issue today. I have a brand color I'd like to use, but it doesn't look the same in Excalidraw.

#3ad0e6 in Canva:
<img width="62" height="76" alt="Image" src="https://github.com/user-attachments/assets/0a9dd559-4f6b-4847-a88e-d87afd2a7e91" />

#3ad0e6 in Firefox:
<img width="170" height="129" alt="Image" src="https://github.com/user-attachments/assets/b2428762-d17f-4e67-880b-e97666a3b5fb" />

#3ad0e6 in Excalidraw:
<img width="91" height="68" alt="Image" src="https://github.com/user-attachments/assets/99470e0c-8718-4158-a2a5-ae6a1d772491" />

Edit:

After re-reading the earlier replies to this thread, I solved by changing to light mode and then adding a large square to serve as the background for my diagram. Works fine, might be worth a tooltip to highlight that the hex color selection is only true in light mode, and it will appear modified in dark mode.

**Comment by snazzyfox** [comment]
Also ran into this issue today on a custom whiteboard. I would love for at least the color processing to be made optional at the app/code level. Right now this behavior appears to be hardcoded, so it's not possible to disable effectively even in an app I control. 

The default invert-rotate setup does keep the colors relatively readable and semantic (in terms of "a little red" vs "very red") for users of both color schemes. But that makes a big assumption that the user is 1) drawing on the board itself, and 2) colors are just used for emphasis/grouping, and the actual colors are unimportant. 

There are many cases where these assumptions aren't true, for example:

- Annotating over, or referencing an image. The annotations can be visible/readable for one person but near invisible for another due to the color shift. Or someone might point to a chart with light/dark colors, only for dark mode to completely flip light/dark.
- Any time the color actually matters, for example, design sketches, where everyone seeing the same color is more important than it being readable.

---

Here's a quick made-up example that doesn't even need color accuracy, that dark mode completely obliterates due to referring to an image:

<details><summary>screenshots</summary>
<p>


<img width="1025" height="518" alt="Image" src="https://github.com/user-attachments/assets/94c58319-4b66-4e33-ae68-22e9a3247968" />

<img width="991" height="445" alt="Image" src="https://github.com/user-attachments/assets/b294e202-8e31-46f0-99cf-862d4f563451" />


</p>
</details> 

What did the author mean to say? It's impossible to tell. It totally depends on which color scheme they had been using when the drawing was made.

---

For anyone who runs into this using excalidraw in their own app, this behavior can be monkey-patched out of the package for the time being like so: 


<details><summary>code</summary>

`src/shims/excalidraw-common.ts`
```ts name=1
 // path depend on your project structure and package manager
export * from "../../node_modules/.pnpm/@excalidraw+common@0.18.0-2b1c2a9/node_modules/@excalidraw/common/dist/prod/index.js";
import * as real from "../../node_modules/.pnpm/@excalidraw+common@0.18.0-2b1c2a9/node_modules/@excalidraw/common/dist/prod/index.js";

// Replace the color handler function with your own which ignores all colors other than grids
export function applyDarkModeFilter(color: string, ...rest: [boolean?]) {
  // Grid color doesn't pass anything in ...rest; all other places do.
  return rest.length === 0 ? real.applyDarkModeFilter(color) : color;
}
```

The app itself unfortunately need to force a background color since auto-flip is also based on the same color processing that makes it lose information, so make sure you disable the picker too:

```ts
<Excalidraw
  onChange={(_, appState) =>{
    const viewBackgroundColor = appState.theme === 'dark' ? '#1b1b1b' : 'transparent';
    if (appState.viewBackgroundColor !== viewBackgroundColor) api.updateScene({
      appState: { viewBackgroundColor },
      captureUpdate: CaptureUpdateAction.NEVER,
    });
  }
  ...
```

Finally make vite load your shim rather than real excalidraw code
```ts
resolve: {
  alias: [
  {
    find: /^@excalidraw\/common$/,
    replacement: fileURLToPath(new URL('./src/shims/excalidraw-common.ts', import.meta.url)),
  },
]},
```

</details>
